### ISSUE TITLE
Rule proposal: no-implicit-conversion

### ISSUE THREAD

**Issue by axetroy** [sindresorhus/eslint-plugin-unicorn#2448]
### Description

When using the +/- operator with strings and numbers.

There are many traps here, and things can go wrong.

e.g.

```js
let result = '5' + 3;
console.log(result); // '53' - String concatenation

result = 3 + '5';
console.log(result); // '35' - Still string concatenation

let result = '10' - 3;
console.log(result); // 7 - The string '10' is converted to the number 10

result = '5' - '2';
console.log(result); // 3 - Both strings are converted to numbers and then subtracted
```

**I know, no one would write code like this in the real world**

But there will be the following situation

```js
function foo() {
  // ...other logic
  return "5"
}

// ... other logic

const bar = foo() + 5 // ❌
```

We should check that left and right expression are of the same type to avoid implicit conversion.

It is feasible when we can infer its type, skip check when found `any`/`unknown`/`never`

### Fail

Plus operator:
```js
const foo = "5" // numberic

// ... other logic

const bar = foo + 5 // ❌
```

```js
function foo() {
  return "5"
}

// ... other logic

const bar = foo() + 5 // ❌
```

Minus operator:

```js
const foo = "10"

// ...other logic

const bar = foo - 5 // ❌
```

Type Annotation:

```js
let foo: string

// ...other logic

const bar = foo - 5 // ❌
```

```js
let foo = (): string => {}

// ...other logic

const bar = foo() - 5 // ❌
```

```js
function foo (bar: string) {
  bar - 5 // ❌
}
```

### Pass

```js
const foo = "5"

// ... other logic

const bar = Number(foo) + 5 // ✅
```

```js
function foo() {
  return "5"
}

// ... other logic

const bar = Number(foo()) + 5 // ✅
```

Minus operator:

```js
const foo = "10"

// ...other logic

const bar = Number(foo) - 5 // ✅
```

### Proposed rule name

no-implicit-conversion

**Comment by fisker** [comment]
This is hard to detect, but 

```js
1e13 - new Date(); // -> number

1e13 + new Date(); // -> string
```

**Comment by fregante** [comment]
If the code is in typescript this is redundant, so the “type annotation” examples shouldn’t be considered. 

For the rest, this would only work in the limited cases where the value (and therefore its type) is defined locally. It kinda makes sense as a rule but also people who care about this just use typescript. By this I mean that outside TS people actually take advantage of JS’ dynamicity and don’t like the verbosity of `Number(str) - 1`. 

The only scenario where this would actually be beneficial (i.e. prevent bugs) is on `str + number`, but it feels like this rarely happens when `str` is declared locally. 

Just my 2c, this rule wouldn’t affect me as I used TS almost exclusively.

**Comment by sindresorhus** [comment]
I think this is better left to typescript-eslint.

For typed projects, `@typescript-eslint/restrict-plus-operands` already covers the core `+` case with real type information. For JavaScript, a useful implementation would quickly become a weak type checker, while a narrow local literal check would catch too little real world code to justify a new Unicorn rule.

ESLint's built in `no-implicit-coercion` also already covers the shorthand coercion style cases like `+foo`, `foo - 0`, `1 * foo`, and `'' + foo`.

Closing as out of scope
